### PR TITLE
ddl: skip adding index reorg work for empty tables

### LINKED ISSUE #51421: flaky test TestDumpPlanReplayerAPI
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 3. What did you see instead (Required)

```
     plan_replayer_test.go:165: 
        	Error Trace:	pkg/server/handler/optimizor/plan_replayer_test.go:165
        	            				pkg/server/handler/optimizor/plan_replayer_test.go:199
        	Error:      	Received unexpected error:
        	            	remove /tmp/plan_replayer.zip: no such file or directory
        	Test:       	TestDumpPlanReplayerAPI 
```

### LINKED ISSUE #54230: add index for small partition table on v8.1.0 is slower than it on v7.5.1
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
drop table if exists tt1, tt2;
set global tidb_partition_prune_mode='dynamic';
set session tidb_partition_prune_mode='dynamic';
CREATE TABLE tt1 (
    id INT NOT NULL,
    listid INT,
    name varchar(10),
    primary key (listid) clustered
)
PARTITION BY LIST (listid) (
    PARTITION p1 VALUES IN (1),
    PARTITION p2 VALUES IN (2),
    PARTITION p3 VALUES IN (3),
    PARTITION p4 VALUES IN (4)
);
CREATE TABLE tt2 (
    id INT NOT NULL,
    listid INT
);
create index idx_listid on tt1(id,listid);
create index idx_listid on tt2(listid);

```
### 3. What did you see instead (Required)
on v7.5.1 cluster
```
mysql> create index idx_listid on tt1(id,listid);
Query OK, 0 rows affected (0.58 sec)

mysql> create index idx_listid on tt2(listid);
Query OK, 0 rows affected (0.51 sec)
```

on v8.1.0 and nightly cluster
```
mysql> create index idx_listid on tt1(id,listid);
Query OK, 0 rows affected (1.99 sec)

mysql> create index idx_listid on tt2(listid);
Query OK, 0 rows affected (1.35 sec)
```
tidb_enable_dist_task is off in v7.5.1, and  on in 8.1.0

### 4. What is your TiDB version? (Required)

v8.1.0 and master

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #54230, close #51421

Problem Summary:

We can reduce the overhead of data reorganization or distributed framework scheduling if the target table is empty.

### What changed and how does it work?

See code changes.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
  Before this PR (empty tables):
  ```sql
  mysql> create index idx_listid on tt1(id,listid);
  Query OK, 0 rows affected (2.52 sec)
  
  mysql> create index idx_listid on tt2(listid);
  Query OK, 0 rows affected (1.52 sec)
  ```
  After this PR (empty tables):
  ```sql
  mysql> create index idx_listid on tt1(id,listid);
  Query OK, 0 rows affected (0.34 sec)
  
  mysql> create index idx_listid on tt2(listid);
  Query OK, 0 rows affected (0.29 sec)
  ```
  Before this PR (non-empty tables):
  ```sql
  mysql> create index idx_listid on tt1(id,listid);
  Query OK, 0 rows affected (2.49 sec)
  
  mysql> create index idx_listid on tt2(listid);
  Query OK, 0 rows affected (2.16 sec)
  ```
  After this PR (non-empty tables):
  ```sql
  mysql> create index idx_listid on tt1(id,listid);
  Query OK, 0 rows affected (2.57 sec)
  
  mysql> create index idx_listid on tt2(listid);
  Query OK, 0 rows affected (2.20 sec)
  ```
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/session/bootstrap.go [hunk 1]
```diff
@@ -3575,7 +3575,7 @@ func rebuildAllPartitionValueMapAndSorted(s *session) {
 			if pi == nil || pi.Type != model.PartitionTypeList {
 				continue
 			}
-			tbl, ok := is.TableByID(s.currentCtx, t.ID)
+			tbl, ok := is.TableByID(context.Background(), t.ID)
```
Comment (Benjamin2037) [inline pkg/session/bootstrap.go:3578]: Why here need update bootstrap part this line code？

**Review by lance6716 — APPROVED** [review]

File: pkg/session/bootstrap.go (hunk 1 above)
Comment (tangenta) [reply pkg/session/bootstrap.go:3578]: During my testing, I triggered several panic cases at this place.

**Review by D3Hunter — APPROVED** [review]
lgtm except existing comments

**Review by Benjamin2037 — APPROVED** [review]
rest LGTM

File: pkg/session/bootstrap.go (hunk 1 above)
Comment (D3Hunter) [reply pkg/session/bootstrap.go:3578]: > I triggered several panic cases at this place.

how is that related to context change?

**Comment by D3Hunter** [comment]
/hold

please pass ctx in bootstrapSessionImpl to rebuildAllPartitionValueMapAndSorted. you can unhold youself after that

**Comment by tangenta** [comment]
> /hold
> 
> please pass ctx in bootstrapSessionImpl to rebuildAllPartitionValueMapAndSorted. you can unhold youself after that

Done.

**Review by yudongusa — APPROVED** [review]
LGTM

**Comment by tangenta** [comment]
/retest

```
go: downloading github.com/pingcap/failpoint v0.0.0-20240527053858-9b3b6e34194a
go: downloading github.com/sergi/go-diff v1.1.0
bazel --output_user_root=/home/jenkins/.tidb/tmp run --config=ci --repository_cache=/share/.cache/bazel-repository-cache //:gazelle
make: execvp: bazel: Permission denied
make: *** [Makefile:591: bazel_ci_simple_prepare] Error 127
./build/jenkins_unit_test.sh: line 26: bazel_collect: command not found
mv: cannot stat 'bazel.xml': No such file or directory
script returned exit code 2
```

**Comment by tangenta** [comment]
/retest

```
    plan_test.go:40: 
        	Error Trace:	pkg/planner/core/casetest/plan_test.go:40
        	            				pkg/planner/core/casetest/tiflash_selection_late_materialization_test.go:76
        	Error:      	Not equal: 
        	            	expected: 4
        	            	actual  : 3
        	Test:       	TestTiFlashLateMaterialization


    plan_stats_test.go:81: 
        	Error Trace:	pkg/planner/core/casetest/planstats/plan_stats_test.go:81
        	            				pkg/planner/core/casetest/planstats/plan_stats_test.go:210
        	Error:      	Not equal: 
        	            	expected: 0
        	            	actual  : -1
        	Test:       	TestPlanStatsLoad
```

**Comment by tangenta** [comment]
/retest

```
    testkit.go:161: 
        	Error Trace:	pkg/testkit/testkit.go:161
        	            				pkg/testkit/testkit.go:154
        	            				pkg/planner/core/casetest/instanceplancache/concurrency_test.go:84
        	            				src/runtime/asm_amd64.s:1700
        	Error:      	Received unexpected error:
        	            	[executor:1213]Deadlock found when trying to get lock; try restarting transaction
        	            	github.com/pingcap/errors.AddStack
        	            		external/com_github_pingcap_errors/errors.go:178
        	            	github.com/pingcap/errors.Trace
        	            		external/com_github_pingcap_errors/juju_adaptor.go:15
        	            	github.com/pingcap/tidb/pkg/testkit.(*TestKit).ExecWithContext
        	            		pkg/testkit/testkit.go:443
        	            	github.com/pingcap/tidb/pkg/testkit.(*TestKit).MustExecWithContext
        	            		pkg/testkit/testkit.go:159
        	            	github.com/pingcap/tidb/pkg/testkit.(*TestKit).MustExec
        	            		pkg/testkit/testkit.go:154
        	            	pkg/planner/core/casetest/instanceplancache/instanceplancache_test.(*worker).run
        	            		pkg/planner/core/casetest/instanceplancache/concurrency_test.go:84
        	            	runtime.goexit
        	            		src/runtime/asm_amd64.s:1700
        	Test:       	TestInstancePlanCacheConcurrencySysbench
        	Messages:   	sql:update normal.sbtest set c='4639' where id=157, [], error stack [executor:1213]Deadlock found when trying to get lock; try restarting transaction
        	            	github.com/pingcap/errors.AddStack
        	            		external/com_github_pingcap_errors/errors.go:178
        	            	github.com/pingcap/errors.Trace
        	            		external/com_github_pingcap_errors/juju_adaptor.go:15
        	            	github.com/pingcap/tidb/pkg/testkit.(*TestKit).ExecWithContext
        	            		pkg/testkit/testkit.go:443
        	            	github.com/pingcap/tidb/pkg/testkit.(*TestKit).MustExecWithContext
        	            		pkg/testkit/testkit.go:159
        	            	github.com/pingcap/tidb/pkg/testkit.(*TestKit).MustExec
        	            		pkg/testkit/testkit.go:154
        	            	pkg/planner/core/casetest/instanceplancache/instanceplancache_test.(*worker).run
        	            		pkg/planner/core/casetest/instanceplancache/concurrency_test.go:84
        	            	runtime.goexit
        	            		src/runtime/asm_amd64.s:1700
```

**Comment by tangenta** [comment]
/retest

```
    queue_v2_ddl_handler_test.go:437: 
        	Error Trace:	pkg/statistics/handle/autoanalyze/priorityqueue/queue_v2_ddl_handler_test.go:437
        	Error:      	Not equal: 
        	            	expected: 114
        	            	actual  : 110
        	Test:       	TestExchangeTablePartition
```

https://do.pingcap.net/jenkins/blue/rest/organizations/jenkins/pipelines/pingcap/pipelines/tidb/pipelines/pull_unit_test_ddlv1/runs/256/nodes/63/steps/69/log/?start=0

```
    distsql_test.go:346: 
        	Error Trace:	pkg/executor/distsql_test.go:346
        	Error:      	"1" is not greater than "2"
        	Test:       	TestCoprocessorPagingSize
```

https://do.pingcap.net/jenkins/blue/rest/organizations/jenkins/pipelines/pingcap/pipelines/tidb/pipelines/ghpr_unit_test/runs/20924/nodes/63/steps/69/log/?start=0

**Comment by tangenta** [comment]
/retest

https://github.com/pingcap/tidb/issues/56797

```
    split_table_test.go:64: 
        	Error Trace:	pkg/executor/test/splittest/split_table_test.go:64
        	Error:      	Expect "t_112_5f7203800000000000000183800000000000" to match "t_115_"
        	Test:       	TestClusterIndexShowTableRegion
```

https://do.pingcap.net/jenkins/blue/rest/organizations/jenkins/pipelines/pingcap/pipelines/tidb/pipelines/ghpr_unit_test/runs/20982/nodes/63/steps/69/log/?start=0

**Pull request merged**